### DISCUSSION TITLE
isolation_level apparently conflicting statements in docs?

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#8252] [Usage Questions]
I'm probably misunderstanding the docs but [Transaction Isolation Level](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html?highlight=dialect#transaction-isolation-level) says:

> Most SQLAlchemy dialects support setting of transaction isolation level using the [create_engine.execution_options](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.execution_options) parameter at the [create_engine()](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine) level

Meanwhile at [Engine Creation API](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine) we find:

> execution_options – Dictionary execution options which will be applied to all connections. See [Connection.execution_options()](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Connection.execution_options)
and
> isolation_level –  this string parameter is interpreted by various dialects in order to affect the transaction isolation level of the database connection. The parameter essentially accepts some subset of these string arguments: "SERIALIZABLE", "REPEATABLE READ", "READ COMMITTED", "READ UNCOMMITTED" and "AUTOCOMMIT". Behavior here varies per backend, and individual dialects should be consulted directly.
> Note that the isolation level can also be set on a per-[Connection](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Connection) basis as well, using the [Connection.execution_options.isolation_level](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Connection.execution_options.params.isolation_level) feature.

I am left quite confused as to what the difference is between passing isolation_level vs exexcution_options=dict(isolation_level=...) to create_engine()? Maybe there is no practical difference but it might be helpful to state this explicitly, or conversely, to give an example as to when the two may be different.

EDIT: Re-reading the explanation I just realized maybe the only difference is that one is allowing you to pass a dictionary vs the other is only allowing you to set the equivalent of the one member of that dictionary that is the `isolation_level` member? If I have that right, I think it would be helpful to state this, and to also clarify which takes precedence if both are passed. 

EDIT2:  No, that was wrong. See below for correct explanation (one sets per statement, one sets when connection is acquired)

I'd be happy to submit a clarifying doc PR for your review if you explain this here and it isn't explained elsewhere.

It's also not clear how AUTOCOMMIT interacts with the other transaction isolation levels because it seems to me that whether or not every statement is auto-committed is an orthogonal issue as to whether the statements within the transaction will yield repeatable read or read uncommitted results etc. This latter question I realize is not SQLA specific per-se, but even looking at postgres docs, the autocommit setting is separate from the transaction isolation; yet here we've got autocommit as an isolation level that exclused selecting a different isolation level together with autocommit.

As always, thanks!

**Comment by zzzeek** [comment]
the docs that get into these different parameters is at https://docs.sqlalchemy.org/en/20/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit .    there is a difference between the straight parameter and the execution option, it's not substantial but it felt important enough that I kept both APIs in.  

so for that, I am horrified that the doc at https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.isolation_level does not have a link to the above link in the seealso's, which should really be the *only* link.    people need to always go to https://docs.sqlalchemy.org/en/20/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit  first - those DBAPI-specific links are a little stale now.

currently, AUTOCOMMIT means there is no "transaction" so there's no transaction isolation level to be configured.    the general idea for that I got from psycopg2.

**Reply by aathan** [reply]
What I can find the the postgres docs (including for example [SET AUTOCOMMIT](https://www.postgresql.org/docs/11/ecpg-sql-set-autocommit.html) doesn't say that autocommit means no transaction. What it says is that it means every statement is its own implicit transaction.

This is why I'm saying "autocommit" is orthogonal to "read commited" or "serialized" etc.

My (possibly totally wrong) impression is that autocommit needs to be separate because you may want autocommit+serialized, for example; and in fact, it appears postgres itself allows you to set those two independently.

In other words, you may not want to wrap each statement in a begin/commit but you still want the results from a given statement to be SERIALIZED results relative to every other statement emitted by your system. Otherwise you might emit two queries spaced milliseconds apart, after some write, and get two different answers.

**Reply by aathan** [reply]
It took me a while to find it but it appears that:

> [create_engine.isolation_level](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.isolation_level) parameter to create_engine():
> ... each new DBAPI connection the moment it’s created will be set to use a "REPEATABLE READ" isolation level setting for all subsequent operations.

but for:

> [create_engine.execution_options](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.execution_options) parameter
> ... the DBAPI connection will be set to use a "REPEATABLE READ" isolation level setting for each new transaction begun

I think those two sentences would be great additions to the documentation of each of those parameters in the create_engine() documentation vs only [here](https://docs.sqlalchemy.org/en/20/core/connections.html#setting-transaction-isolation-levels-including-dbapi-autocommit).

**Reply by zzzeek** [reply]
> What I can find the the postgres docs (including for example [SET AUTOCOMMIT](https://www.postgresql.org/docs/11/ecpg-sql-set-autocommit.html) doesn't say that autocommit means no transaction. What it says is that it means every statement is its own implicit transaction.
> 
> This is why I'm saying "autocommit" is orthogonal to "read commited" or "serialized" etc.

yes we are aware of this, see #6489, #3357.  however we dont yet have this concept expressed and it's also very difficult to find a case that it matters in practice, especially on PostgreSQL (if that's the target DB we're discussing).   

For PG, it's hard to see how it would matter because the driver (like psycopg2, etc. ) does not emit BEGIN when we are in autocommit mode.   then, per https://www.postgresql.org/docs/current/sql-set-transaction.html "If SET TRANSACTION is executed without a prior START TRANSACTION or BEGIN, it emits a warning and otherwise has no effect." .   so arguably this seems to suggest isolation mode on PG is not in effect when BEGIN is not emitted.

> 
> My (possibly totally wrong) impression is that autocommit needs to be separate because you may want autocommit+serialized, for example; and in fact, it appears postgres itself allows you to set those two independently.

sure, we have #6489 open, but this is one of those issues that usually just hang on forever because in practice, nobody needs this.   

> In other words, you may not want to wrap each statement in a begin/commit but you still want the results from a given statement to be SERIALIZED results relative to every other statement emitted by your system. Otherwise you might emit two queries spaced milliseconds apart, after some write, and get two different answers.

sure, this is demonstrated at https://stackoverflow.com/questions/24706867/whats-the-point-of-using-high-isolation-level-when-in-autocommit-mode but on PostgreSQL it's not clear how to actually effect this, if I'm reading their statement correctly, since it seems to imply that if dont call BEGIN, there's no way to actually effect the isolation level.

this is super in the weeds stuff.   SERIALIZABLE is already a seldom used isolation level for it's impracticability, and then within the realm of not using BEGIN at all, if someone truly has this issue they can emit the isolation level commands they need (which are unclear at the moment) directly.

**Reply by zzzeek** [reply]
i would add if you want to get us better information for #6489 particularly this PG behavior, we can add a note to the existing docs indicating that this is not explicitly supported right now and how to work around.

**Reply by zzzeek** [reply]
I'm correcting the PG docs to be consitent with the others which refer to the isolation_level parameter of create_engine directly, which should likely be preferred for engine-wide isolation level settings.

**Reply by aathan** [reply]
Yes, I was going to go interact with PG maintainers and see what they have to say about it. I'll revert back with some info for ya.

**Reply by aathan** [reply]
By the way, when I said SERIALIZED (SERIALIZABLE) above it was just an example. Replace with READ COMMITTED if you want. I have specifically witnessed what I believe was that under certain kinds of load conditions, an autocommit session will end up returning different results on closely spaced queries and I would have liked to be able to be in DBAPI autocommit but SERALIZED isolation. The same issue was visible with non DBAPI autocommit sessions ... i.e., the result behavior is due to the isolation level, and was independent of the autocommit setting.
